optaplanner-core/src/main/java/org/optaplanner/core/impl/phase/AbstractPhase.java

```diff
@@ -129,16 +113,21 @@ public void solvingEnded(SolverScope<Solution_> solverScope) {
     public void phaseStarted(AbstractPhaseScope<Solution_> phaseScope) {
         phaseScope.startingNow();
         phaseScope.reset();
-        bestSolutionRecaller.phaseStarted(phaseScope);
-        solverPhaseLifecycleSupport.firePhaseStarted(phaseScope);
+        phaseScope.getSolverScope().getSolver().phaseStarted(phaseScope);
         termination.phaseStarted(phaseScope);
         phaseLifecycleSupport.firePhaseStarted(phaseScope);
     }
 
+    @Override
+    public void phaseEnded(AbstractPhaseScope<Solution_> phaseScope) {
+        phaseScope.getSolverScope().getSolver().phaseEnded(phaseScope);
+        termination.phaseEnded(phaseScope);
+        phaseLifecycleSupport.firePhaseEnded(phaseScope);
```

Comment:
(important) Shouldn't this be inside-out? so reversed order as in the original code.

Or maybe not? The phase lifecycle vents might want to use the solving end time. Other the hand, the solving end time might need to include the execution of the ending events?

Take a look at other phaseStarted vs phaseEnded implementations. Does the end phase do things in the same order or in the reverse order of start events?

Something to discuss! We need a general guideline for this.